### PR TITLE
1/n Add `FaultTolerantMode`

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Find the complete PoC here: https://github.com/PyTorchLightning/pytorch-lightning/pull/10568

Parts of #10431

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/utilities/enums.py [hunk 1]
```diff
@@ -256,3 +256,22 @@ def interactive_compatible_types() -> List["_StrategyType"]:
     def is_interactive_compatible(self) -> bool:
         """Returns whether self is interactive compatible."""
         return self in _StrategyType.interactive_compatible_types()
+
+
+class FaultTolerantTrainingMode(LightningEnum):
```
Comment (ananthsub) [inline pytorch_lightning/utilities/enums.py:261]:
```suggestion
class _FaultTolerantTrainingMode(LightningEnum):
```

File: pytorch_lightning/utilities/enums.py [hunk 2]
```diff
@@ -256,3 +256,22 @@ def interactive_compatible_types() -> List["_StrategyType"]:
...
+    def is_enabled(self) -> bool:
+        return self is not FaultTolerantTrainingMode.DISABLED
+
+    @property
+    def is_automatic(self) -> bool:
+        return self is not FaultTolerantTrainingMode.AUTOMATIC
+
+    @property
+    def is_manual(self) -> bool:
+        return self is not FaultTolerantTrainingMode.MANUAL
```
Comment (ananthsub) [inline pytorch_lightning/utilities/enums.py:277]: should these be `self is <>` instead of `self is not <>` ?

Comment (tchaton) [reply pytorch_lightning/utilities/enums.py]: Done !

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -571,3 +571,19 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
...
+        return _FaultTolerantTrainingMode.DISABLED
+    elif env_value == "1":
+        return _FaultTolerantTrainingMode.AUTOMATIC
+    elif env_value == "2":
+        return _FaultTolerantTrainingMode.MANUAL
+    else:
+        raise MisconfigurationException(
+            "The environnement flag `PL_FAULT_TOLERANT_TRAINING` should be either "
+            "'0' (disabled), '1' (automatic) or '2' (manual)."
+        )
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:589]:
```suggestion
    raise MisconfigurationException(
        "The environnement flag `PL_FAULT_TOLERANT_TRAINING` should be either "
        "'0' (disabled), '1' (automatic) or '2' (manual)."
    )
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -571,3 +571,19 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
...
+
+
+def _detect_fault_tolerant_env_to_enum() -> _FaultTolerantTrainingMode:
+    """This utility detect if Fault Tolerant is activated and maps its value to `_FaultTolerantTrainingMode`."""
+    env_value = os.getenv("PL_FAULT_TOLERANT_TRAINING", "0")
+    if env_value == "0":
+        return _FaultTolerantTrainingMode.DISABLED
+    elif env_value == "1":
+        return _FaultTolerantTrainingMode.AUTOMATIC
+    elif env_value == "2":
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:583]: Neat! The settings feel like a gam 😄

Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:583]: Neat! The settings feel like a game 😄

File: pytorch_lightning/utilities/auto_restart.py [hunk 5]
```diff
@@ -571,3 +571,19 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _detect_fault_tolerant_env_to_enum() -> _FaultTolerantTrainingMode:
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:576]:
```suggestion
def _detect_fault_tolerant_training_mode() -> _FaultTolerantTrainingMode:
```

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:589]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py]: Done !

File: pytorch_lightning/utilities/auto_restart.py [hunk 6]
```diff
@@ -571,3 +571,18 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _detect_fault_tolerant_training_mode() -> _FaultTolerantTrainingMode:
+    """This utility detect if Fault Tolerant is activated and maps its value to `_FaultTolerantTrainingMode`."""
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:577]:
```suggestion
    """This utility detects if Fault Tolerant is activated and maps its value to `_FaultTolerantTrainingMode`."""
```

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 7]
```diff
@@ -571,3 +571,18 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
...
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _detect_fault_tolerant_training_mode() -> _FaultTolerantTrainingMode:
+    """This utility detects if Fault Tolerant is activated and maps its value to `_FaultTolerantTrainingMode`."""
+    env_value = os.getenv("PL_FAULT_TOLERANT_TRAINING", "0")
+    if env_value == "0":
+        return _FaultTolerantTrainingMode.DISABLED
+    elif env_value == "1":
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:581]: Should we do something like

```suggestion
    if env_value in ("0", "disabled"):  # 0 kept for compatibility
        return _FaultTolerantTrainingMode.DISABLED
    elif env_value == "automatic":
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 8]
```diff
@@ -571,3 +571,18 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _detect_fault_tolerant_training_mode() -> _FaultTolerantTrainingMode:
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:576]: This should be a classmethod in `_FaultTolerantTrainingMode`

File: pytorch_lightning/utilities/auto_restart.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:581]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:576]: Done !

**Review by carmocca — APPROVED** [review]
